File: solutions/ruby/kindergarten-garden/1/kindergarten_garden.rb
# frozen_string_literal: true

# class Garden
class Garden
  STUDENTS = %i[alice bob charlie david eve fred ginny harriet ileana joseph kincaid larry].freeze
  STUDENTS.each_with_index { |student, index| define_method(student) { plants_for(index) } }
  def initialize(diagram) = @rows = diagram.lines(chomp: true)

  private

  PLANTS = { 'C' => :clover, 'G' => :grass, 'R' => :radishes, 'V' => :violets }.freeze
  def plants_for(index) = @rows.flat_map { |row| row[index * 2, 2].chars.map(&PLANTS) }
end

File: solutions/clojure/kindergarten-garden/1/src/kindergarten_garden.clj
(ns kindergarten-garden)

(def plants {\C :clover \G :grass \R :radishes \V :violets})
(defn garden [s]
  (let [[_ row1 row2] (re-matches #"(\w+)\n(\w+)" s)
        top_row (mapv plants row1)
        bottom_row (mapv plants row2)]
    (into {} (map #(identity [%1 (concat %2 %3)])
                  [:alice :bob :charlie :david :eve :fred :ginny
                   :harriet :ileana :joseph :kincaid :larry]
                  (partition 2 top_row)
                  (partition 2 bottom_row)))))
